### PR TITLE
RES 1647 - Online EWC mixin

### PR DESCRIPTION
**Pull request by lucasosouza**
Minimal implementation of Online EWC, based on https://github.com/kuc2477/pytorch-ewc. Overall weights, gradients and updated were validated and match original implementation. 

Missing from this implementation, to include in a following PR:
- Adaptation of Continual Learning to run the domain scenario (may require another subclass of Experiment, but I'm looking for a way to do it without). The original repo only tests the implementation in the domain scenario (Permuted MNIST), making it harder to perform an exact comparison.
- Calculate diagonal of the Fisher information matrix in the distributed setting, taking advantage of multiple GPUs/nodes.

### CODE REVIEW

File: nupic/research/frameworks/pytorch/models/cl_models.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            nn.Linear(self.hidden_size, self.output_size)
+        ])
+
+        self.xavier_initialize()
+
+    @property
+    def name(self):
+        return (
+            "MLP"
+            "-in{input_size}-out{output_size}"
```
[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/research/frameworks/pytorch/models/cl_models.py:70]: Maybe this is served better as `extra_repr`.

File: nupic/research/frameworks/pytorch/models/cl_models.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            hidden_layer_num=self.hidden_layer_num,
+            input_dropout_prob=self.input_dropout_prob,
+            hidden_dropout_prob=self.hidden_dropout_prob,
+        )
+
+    def forward(self, x):
+        return reduce(lambda x, l: l(x), self.layers, x)
+
+    def xavier_initialize(self):
+        modules = [
```
Comment (mvacaporale) [inline nupic/research/frameworks/pytorch/models/cl_models.py:86]: I'm not opposed to this format, but the string comprehensions are a bit long here. Maybe try:
```
for m in self.modules():
     if not isinstance(m, (nn.Linear, nn.Conv2d)):
          continue
     for p in m.parameters():
          if p.dim() >= 2:
               init.xavier_normal(p)
```

File: nupic/research/frameworks/vernon/mixins/ewc.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+import torch
+import torch.nn.functional as F
+from torch import autograd
+
+
+class ElasticWeightConsolidation:
+    """
+    Online version of Elastic Weight Consolidation.
+    Diagonals of the Fisher information matrix is calculated at the end
```
Comment (mvacaporale) [inline nupic/research/frameworks/vernon/mixins/ewc.py:30]: "are calculated"

File: nupic/research/frameworks/vernon/mixins/ewc.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+                F.log_softmax(self.model(x))[range(self.batch_size), y.data]
+            )
+            # Can't use the full dataset, too expensive
+            if idx >= self.ewc_fisher_num_batches:
+                break
+
+        # Convert into list of tensors per sample
+        loglikelihoods = torch.cat(loglikelihoods).unbind()
+        # Estimate the fisher information of the parameters.
+        grads = zip(*[autograd.grad(
```
Comment (mvacaporale) [inline nupic/research/frameworks/vernon/mixins/ewc.py:86]: These next few lines are a bit busy. It may help to break it down for readability.

**Review by mvacaporale — APPROVED** [review]

File: projects/continual/experiments/__init__.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
+# ----------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply projects/continual/experiments/__init__.py:1]: Good catch. There was a confusion with the other PR that was about to merge, I didn't want to create merge conflicts. I will fix it now it has merged

File: nupic/research/frameworks/pytorch/models/cl_models.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+class EWCNetwork(nn.Module):
+
+    def __init__(self, input_size, output_size,
+                 hidden_size=400,
+                 hidden_layer_num=2,
+                 hidden_dropout_prob=.5,
+                 input_dropout_prob=.2,
+                 ):
+        """
+        Adapted as is from
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/models/cl_models.py:36]: As is means "exactly as it is in the original repo". How should I rephrase it?

File: nupic/research/frameworks/pytorch/models/cl_models.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        https://github.com/kuc2477/pytorch-ewc/blob/master/model.py
+
+        Used to test consistency in the algorithm implementation.
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/models/cl_models.py:39]: It replicates a model from the repo, not necessarily the same as the one in the paper. It is here in order to be able to compare it to that repo.

File: nupic/research/frameworks/pytorch/models/cl_models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/models/cl_models.py:70]: This network is just copied as it is, and not intended to be used except for the sole purpose of validating the implementation. Hence I didn't make any changes

File: nupic/research/frameworks/pytorch/models/cl_models.py (hunk 2 above)
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/models/cl_models.py:86]: Same as above

File: nupic/research/frameworks/pytorch/models/cl_models.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/research/frameworks/pytorch/models/cl_models.py:36]: How about, "Adapted nearly verbatim from"

File: nupic/research/frameworks/pytorch/models/cl_models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/models/cl_models.py:70]: Maybe this should be in the projects folder instead of frameworks, given it is only for that specific project? What do you think?

File: nupic/research/frameworks/pytorch/models/cl_models.py (hunk 2 above)
Comment (mvacaporale) [reply nupic/research/frameworks/pytorch/models/cl_models.py:86]: Nevermind about this, I don't think mine is any better.

File: nupic/research/frameworks/pytorch/models/cl_models.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/research/frameworks/pytorch/models/cl_models.py:39]: I agree with the strategy, but I'm hesitant to incorporate a network in `frameworks` unless its implementation is polished and maximally readable and maintainable. The `name` property doesn't affect the functionality of the forward or backwards pass, so I suggest we clean it up. Another possibility is to just move this network to the project folder.

[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/research/frameworks/pytorch/models/cl_models.py:39]: Gotcha

File: nupic/research/frameworks/pytorch/models/cl_models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/research/frameworks/pytorch/models/cl_models.py:70]: I agree with the strategy, but I'm hesitant to incorporate a network in `frameworks` unless its implementation is polished and maximally readable and maintainable.

The `name` property doesn't affect the functionality of the forward or backwards pass, so it's not too risky to clean it up. Another possibility is to just move this network to the project folder.

[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/research/frameworks/pytorch/models/cl_models.py:70]: There's a general question to ask when adding something to frameworks, "will this most likely be used in other projects". If the answer is yes, then I think we should keep the implementation as clean and general as possible.
